### PR TITLE
[SPARK-35906][SQL] Remove order by if the maximum number of rows less than or equal to 1

### PR DESCRIPTION
**Pull request by wangyum**
### What changes were proposed in this pull request?

This PR removes order by if the maximum number of rows less than or equal to 1. For example:
```scala
spark.sql("select count(*) from range(1, 10, 2, 2) order by 1 limit 10").explain("cost")
```
Before this pr:
```
== Optimized Logical Plan ==
Sort [count(1)#2L ASC NULLS FIRST], true, Statistics(sizeInBytes=16.0 B)
+- Aggregate [count(1) AS count(1)#2L], Statistics(sizeInBytes=16.0 B, rowCount=1)
   +- Project, Statistics(sizeInBytes=20.0 B)
      +- Range (1, 10, step=2, splits=Some(2)), Statistics(sizeInBytes=40.0 B, rowCount=5)
```

After this pr:
```
== Optimized Logical Plan ==
Aggregate [count(1) AS count(1)#2L], Statistics(sizeInBytes=16.0 B, rowCount=1)
+- Project, Statistics(sizeInBytes=20.0 B)
   +- Range (1, 10, step=2, splits=Some(2)), Statistics(sizeInBytes=40.0 B, rowCount=5)
```

### Why are the changes needed?

Improve query performance.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Unit test.

### CODE REVIEW

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM (Pending CIs).
I checked the all revised queries. The plan makes sense. Thanks, @wangyum .

**Comment by dongjoon-hyun** [comment]
Could you check the UT failure?
```
SPARK-24556: always rewrite output partitioning in ReusedExchangeExec and InMemoryTableScanExec *** FAILED *** (40 milliseconds)
ArrayBuffer() had size 0 instead of expected size 1 (PlannerSuite.scala:746)
```

**Comment by dongjoon-hyun** [comment]
Retest this please

**Comment by dongjoon-hyun** [comment]
It seems that Jenkins is down. Could you re-trigger your GitHub Action?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala [hunk 1]
```diff
@@ -1218,6 +1219,7 @@ object EliminateSorts extends Rule[LogicalPlan] {
     _.containsPattern(SORT))(applyLocally)
 
   private val applyLocally: PartialFunction[LogicalPlan, LogicalPlan] = {
+    case Sort(_, _, child) if child.maxRows.exists(_ <= 1L) => child
```
Comment (HyukjinKwon) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:1222]: The change here looks good. I wonder if we should better create a separate file to group similar stuff though .... e.g.) we can remove deduplicate too.

**Review by HyukjinKwon — APPROVED** [review]
Looks good as is. just one thought.

**Review by maropu — APPROVED** [review]

**Review by huaxingao — APPROVED** [review]

**Comment by dongjoon-hyun** [comment]
Thank you, @wangyum , @HyukjinKwon , @maropu , @huaxingao .
For the above comment (https://github.com/apache/spark/pull/33100#discussion_r659455547), we can discuss later since it's about refactoring.

Merged to master for Apache Spark 3.2.0.

**Comment by dongjoon-hyun** [comment]
cc @cloud-fan

**Comment by cloud-fan** [comment]
late LGTM

**Comment by allisonwang-db** [comment]
Hi @wangyum It seems that the `maxRow` check should be added in `recursiveRemoveSort` instead of `applyLocally`. Current the Once strategy's idempotency can be broken if the plan has two consecutive sorts that are removable:
```scala
Sort [a ASC]
+- Sort [a ASC]
```

Example query:
```sql
select a from (select a from values(0, 1) t(a, b) order by a) order by a
```

**Comment by wangyum** [comment]
Thank you @allisonwang-db.

**Comment by HyukjinKwon** [comment]
Should we revert this, @wangyum and @allisonwang-db ?

**Comment by wangyum** [comment]
I have a PR to fix it: https://github.com/apache/spark/pull/33240

**Pull request closed without merging**